fs/nova/balloc.c

```diff
@@ -601,65 +612,142 @@ static long nova_alloc_blocks_in_free_list(struct super_block *sb,
 	else
 		temp = &(free_list->last_node->node);
 
-	while (temp) {
-		step++;
-		curr = container_of(temp, struct nova_range_node, node);
+	/* Try huge block allocation for data blocks first */
+	if (IS_DATABLOCKS_2MB_ALIGNED(num_blocks, atype)) {
+		unsigned int left_margin;
+		unsigned int right_margin;
+
+		while (temp) {
+			step++;
+			curr = container_of(temp, struct nova_range_node, node);
 
-		if (!nova_range_node_checksum_ok(curr)) {
-			nova_err(sb, "%s curr failed\n", __func__);
-			goto next;
+			if (!nova_range_node_checksum_ok(curr)) {
+				nova_err(sb, "%s curr failed\n", __func__);
+				goto next;
+			}
+
+			curr_blocks = curr->range_high - curr->range_low + 1;
+			left_margin = PAGES_PER_2MB - (curr->range_low & PAGES_PER_2MB_MASK);
+			right_margin = curr_blocks - left_margin - num_blocks;
+
+			/* do the allocaiton once remaining block is large engough*/
+			if (num_blocks <= (curr_blocks - left_margin)) {
+				struct nova_range_node *node;
+
+				nova_dbgv("curr:%p: num_blocks:%lu curr->range_low:%lu high:%lu", curr, num_blocks, curr->range_low, curr->range_high);
+				/* Free curr before insert the left,right margin node. */
```

Comment:
Why? We can reuse curr.

Reply:
w/o remove original node, the following insert would fail at
https://github.com/NVSL/linux-nova/blob/master/fs/nova/balloc.c#L275
as the new node is overlapped with existing one.

https://github.com/NVSL/linux-nova/blob/master/fs/nova/balloc.c#L200
compare node will return 0 in problematic case.